### PR TITLE
Schema

### PR DESCRIPTION
**Pull request by trvrb**
This PR is for discussion and refinement of schema proposals.

### CODE REVIEW

**Comment by rneher** [comment]
I am generally in favor of the relational schema (strain, sample, sequence) -- mostly because of the many samples and sequences etc that could be attached to a strain and I fear that it will bloat otherwise. I don't like the key `accessions`. Plant ppl use it for isolates. I would rather call it `sequences`. Then we probably want a generic phenotype field for drug susceptibilities, titers, etc. What about collection host info in a dict as 
`host:{age:21, species:human, sex:female, status:deceased..., }`.

**Comment by trvrb** [comment]
@rneher ---

Thanks for the thoughts. I think I'm less worried about bloat because Rethink provides an elegant way to update just a subfield in a document. But point taken and worth considering. You can also `pluck` fields when getting a document to avoid downloading the whole thing.

Phenotype and titer data is an important consideration. In the hierarchical model I would think that the most natural thing would be to nest `phenotypes` and `titers` under `sample`. In the relational model these would be two more tables keyed off of `sample`. I imagine these as different things because `phenotype` is an attribute of a single entity while `titer` is an attribute of two entities. I'll spec out `titers` in the flu example just to give this more definition.

I think the `host` dictionary is probably a good idea. However, I would make it a nested dictionary like you say in the nested schema, while I would make it a new table in the relational schema. Otherwise it feels too much like mixing metaphors.

**Comment by rneher** [comment]
re sequences as seperate table: just seems more natural to me given that most data bases we draw from work that way.

re host: I don't mind mixing that much. since strain--host is basically one-to-one, I think keeping a nested dict even when using multiple tables/schemata for strains, samples, sequences, and authors is not that terrible. Main reason being that the other relations are one-to-many while host-strain is one-to-one. separate table seems weird in the that case. 

titers are a funny type of data since the relate many other entries -> separate table

but there are many simpler phenotypes that could just be nested in samples.

in anycase. I am don't want to be too dogmatic about any of these

**Comment by trvrb** [comment]
Cool. This is helpful. It sounds like you agree however that the simpler two-table `strain` / `sequence` schema is not expressive enough? I really didn't like how this performed with multiple samples from the same infection (as in the both the Zika and the flu examples).

**Comment by trvrb** [comment]
Okay, I did some research on performance and document size in RethinkDB. Here's what I found: the is a pretty significant write performance issue when modifying fields in a large document.

> [Stackoverflow](https://stackoverflow.com/questions/19031597/can-rethinkdb-hold-more-than-a-few-hundred-embedded-arrays): The problem with having a large nested array in a document (really just a large document in general there's nothing special about arrays) is that it makes it slow if you need to update it. RethinkDB doesn't do partial updates right now so anytime you want to update the document it will require reading the entire thing of disk and writing the entire thing to disk.

This issue is not resolved and it looks like it will take a while for it to be resolved:
* https://github.com/rethinkdb/rethinkdb/issues/5369
* https://github.com/rethinkdb/rethinkdb/issues/2946

So, if I understand correctly, even though you can make this sort of update in a single ReQL call (although [this is rather messy](https://stackoverflow.com/questions/30265313/rethinkdb-updating-nested-array)), it will still require a large operation on the server.

I think this pushes things pretty strongly in the relational direction for implementation reasons.

**Comment by trvrb** [comment]
These three schemas now have example virus specific fields like `microcephaly_status` and `subtype`. These follow the core fields. The schema in `schema_minimal.json` is the current take on fields that should always be present and explicitly set to `null` if not given in input. You can see the only required fields are `strain_name`, `owner` and `sequence` when adding a sequence record.

Strain id, sample id and sequence id must be unique across each table (and I think better if globally unique within a db / pathogen). Here, I created the sample id as `strain_name|sample_name` and the sequence id as `strain_name|sample_name|sequence_accession`. If a `sample_name` or `sequence_accession` isn't provided it defaults to `1`. All this is necessary to dedup multiple uploads of the same FASTA.

I went with `host_sex` and `host_age` rather than making `host` a dictionary because I wanted to explicitly and directly key these fields off of named FASTA fields. Every field name is unique across the schema to accommodate this.

**Comment by trvrb** [comment]
I wrote a testing script to compare read and write performance in RethinkDB for the relational and nested data models. In doing so, I learned more Rethink syntax. This made it pretty obvious that the flat data model is easier to work with. Here's an update to a sequence attribute in the flat model:

```
r.db('relational').table('sequences').get('USVI/1/2016|VI1.1|1')
  .update({"sequence_locus": "NEWVALUE"}).run(conn)
```

Here is the exact same update with the nested model:

```
r.db('nested').table('strains')
  .get('USVI/1/2016')
  .update({
      'samples': {
          'USVI/1/2016|VI1.1': {
              'sequences': {
                  'USVI/1/2016|VI1.1|1' : {'sequence_locus': "NEWVALUE}
              }
          }
      }
  })
  .run(conn)
```

Here's getting all `sequence_locus` fields associated with a single strain in relational form:

```
r.db('relational').table('strains').get_all('USVI/1/2016').eq_join(
    'id',
    r.db('relational').table('sequences'),
    index='strain_name'
  ).zip()['sequence_locus'].run(conn)
```

Here's getting the same information in nested form:

```
r.db('nested').table('strains').get('USVI/1/2016').do(
  lambda strain: strain['samples'].values().map(
    lambda sample: sample['sequences'].values().map(
        lambda sequence: sequence['sequence_locus']
      )
    )
  ).run(conn)
```


With 1k strains / 10k samples / 100k sequences and 20kb sequence length (Ebola) I get the following performance:

```
Update sample attribute ct
relational updates per second: 900.1
nested updates per second: 841.0
Update sequence attribute sequence_locus
relational updates per second: 849.4
nested updates per second: 771.3
Testing read performance
Get all sequence attribute sequence_locus associated with strain USVI/1/2016
relational reads per second: 629.2
nested reads per second: 940.9
```

With 1k strains / 10k samples / 100k sequences and 1Mb sequence length (tuberculosis) I get the following performance:

```
Update sample attribute ct
relational updates per second: 935.3
nested updates per second: 76.0
Update sequence attribute sequence_locus
relational updates per second: 134.2
nested updates per second: 66.8
Testing read performance
Get all sequence attribute sequence_locus associated with strain USVI/1/2016
relational reads per second: 173.1
nested reads per second: 159.5
```

File: schema/schema_zika.json [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    {
+      "id" : "USVI/1/2016",
+      "strain_name" : "USVI/1/2016",
+      "owner" : "bpotter",
+      "host_species" : "human",
+      "host_age": "23y",
+      "host_sex" : "female",
+      "symptom_onset_date" : "2016-09-25",
+      "symptoms" : null,
+      "pregnancy_status" : true,
```
Comment (jameshadfield) [inline schema/schema_zika.json]: A lot of these (`pregnancy_status`, `pregnancy_week`, `host_age`, `symptoms`) will / may change depending on the sampling date. Should they be in that table instead?

File: schema/schema_zika.json [hunk 2]
```diff
@@ -0,0 +1,180 @@
+{
+  "dbinfo" : {
+    "pathogen" : "zika"
+  },
+  "strains" : [
+    {
+      "id" : "USVI/1/2016",
```
Comment (jameshadfield) [inline schema/schema_zika.json]: shall we rename `id` as `strain_id` or `strain_uid` (the primary key) and have `sacra` make it (it's identical to `strain_name`...). Alternatively, we could skip this and use `strain_name` as the primary key. (The downside being potential confusion -- `sample_name` is not the primary key of the `samples` table)

File: schema/schema_zika.json [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      "collection_date" : "2016-09-28",
+      "country" : "usvi",
+      "division" : "saint_croix",
+      "subdivision" : null,
+      "gps" : null,
+      "collecting_lab" : "USVI DoH",
+      "passage" : "unpassaged",
+      "tissue" : "serum",
+      "ct" : 25,
+      "usvi_doh_sample_id" : "G125"
```
Comment (jameshadfield) [inline schema/schema_zika.json]: having fields such as this (only applicable to a small subset of the data) makes the table really wide. Not a problem, just something to keep in mind.

File: schema/schema_zika.json [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      "host_age": null,
+      "host_sex" : null,
+      "symptom_onset_date" : null,
+      "symptoms" : null,
+      "infection_origin_country" : "dominican_republic"
+    }
+  ],
+  "samples" : [
+    {
+      "id" : "USVI/1/2016|VI1.1",
```
Comment (jameshadfield) [inline schema/schema_zika.json]: primary key: `sample_id` or `sample_uid` - as per the strains table, get sacra to make this out of `strain_name + "|"  + `sample_name`. How sold are we on `|` as the joining symbol?

File: schema/schema_zika.json [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+      "gps" : null,
+      "collecting_lab" : null,
+      "passage" : "cell culture",
+      "tissue" : null,
+      "ct" : null
+    }
+  ],
+  "sequences" : [
+    {
+      "id" : "USVI/1/2016|VI1.1|1",
```
Comment (jameshadfield) [inline schema/schema_zika.json]: as per above, this should be `sequence_id` or `sequence_uid`

**Comment by barneypotter24** [comment]
I've just posted a new dictionary in the Sacra [config](https://github.com/nextstrain/sacra/blob/cleaning_functions/src/cfg.py#L21) that should contain the maximal set of fields in each table. Each field should be unique as @jameshadfield described above and we discussed to day. So far I have lifted from `schema_zika.json`, but I know that there are other fields that will be necessary, so I want to leave a note here to update that dictionary with new fields as we come to realize them.

File: schema/schema_zika.json [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+      "strain_name" : "USVI/1/2016",
+      "sample_name" : "VI1.1",
+      "sequence_accession": "1",
```
Comment (jameshadfield) [inline schema/schema_zika.json]: I'm assuming that if (for example) this now goes to NCBI (new sequence url, accession etc), then we will get a new row in this table? If so, we'll need a method by which to "choose" which sequence to use (e.g. the `choose_genbank` method in fauna)

File: schema/schema_zika.json [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+      "sequencing_lab" : null,
+      "sharing" : "public",
+      "sequence_url" : "https://www.ncbi.nlm.nih.gov/nuccore/KU853013",
+      "attribution" : "barzon_2016_isolation",
+      "sequence" : "gattag..."
+    }
+  ],
+  "attributions" : [
+    {
+      "id" : "black_2017_genetic",
```
Comment (jameshadfield) [inline schema/schema_zika.json]: this looks like a great way to make the primary key 👍 

as per above, `id` -> `attribution_(u)id`

**Comment by trvrb** [comment]
I didn't realize this was never merged. Looks like no conflicts appeared in the ensuing months. Merging now.

**Pull request merged**